File: src/scenario/ui/_htmlgendivs.py
# -*- coding: utf-8 -*-

# Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
HTML div generator.
"""

import typing

import scenario

if True:
    from ._htmlgentypes import CollapsibleState as _CollapsibleStateImpl  # @default-parameter-value
if typing.TYPE_CHECKING:
    from ._htmldoc import HtmlDocument as _HtmlDocumentType
    from ._htmlgentypes import CollapsibleState as _CollapsibleStateType


class DivGenerator:
    """
    HTML generator for a ``<div></div>`` section.

    Optional features:

    - Section name.
    - Collapsible content.

        .. note:: :meth:`addtogglebutton()`/:meth:`addcollapsiblediv()`/:meth:`setstate()` can be used without prior call to :meth:`adddiv()`.
    """

    def __init__(
            self,
            html,  # type: _HtmlDocumentType
            *,
            name="",  # type: str
            collapsible_cid="",  # type: str
            default_state=_CollapsibleStateImpl.EXPANDED,  # type: typing.Optional[_CollapsibleStateType]
    ):  # type: (...) -> None
        """
        Instantiates a :class:`DivGenerator` with configurations.

        :param html:
            HTML output page to feed.
        :param name:
            Optional section name.
        :param collapsible_cid:
            Optional CID for the collapsible content.
        :param default_state:
            Default state for the collapsible content, when ``collapsible_cid`` is defined.

            Default is `expanded`.

            When ``None``, may be set with :meth:`setstate()` in the end.
        """
        from ._htmlgenbuttons import ButtonGenerator

        #: HTML output page to feed.
        self.html = html  # type: _HtmlDocumentType

        #: Optional section name.
        self.name = name  # type: str

        #: Optional collapsible CID.
        self.collapsible_cid = collapsible_cid  # type: str

        #: Default state for the collapsible content.
        #:
        #: Not expandable/callapsible if ``None`` in the end.
        self.default_state = default_state  # type: typing.Optional[_CollapsibleStateType]

        #: Main ``<div></div>`` node context.
        self._main_div_ctx = None  # type: typing.Optional[_HtmlDocumentType.NodeContext]
        #: ``.toggle-div`` button generator.
        self._button_generator = ButtonGenerator(
            html,
            toggle_type="div" if collapsible_cid else "",
            toggle_cid=collapsible_cid,
        )  # type: ButtonGenerator
        #: ``div.collapsible`` node context.
        self._collapsible_div_ctx = None  # type: typing.Optional[_HtmlDocumentType.NodeContext]

        # Have the related .js content be embedded at the end of the HTML page.
        self.html.addfinaljs(scenario.Path(__file__).with_suffix(".js"))

    def __repr__(self):  # type: () -> str
        """
        Canonical string representation.

        For debugging purpose.
        """
        return "DivGenerator(%s)" % (", ".join([
            f"name={self.name!r}",
            f"collapsible_cid={self.collapsible_cid!r}",
        ]))

    def adddiv(
            self,
            *,
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> _HtmlDocumentType.NodeContext
        """
        Adds the main ``<div></div>``.

        :param classes: Optional additional HTML classes.
        :return: HTML node context focused on the ``<div></div>`` node created.
        """
        # Create the main div.
        self._main_div_ctx = self.html.addnode(
            "div",
            classes=[
                *classes,
                "named" if self.name else "",
            ],
        )

        # Initiate the content.
        with self._main_div_ctx:
            # In case of collapsible content, add a toggle button.
            if self.collapsible_cid:
                self.addtogglebutton()

            # In case of a named section, add the name in a `span.name` element.
            if self.name:
                self.html.addnode(
                    "span",
                    classes=[
                        *classes,
                        "name",
                    ],
                    text=self.name,
                )

        return self._main_div_ctx

    def addtogglebutton(
            self,
            *,
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> None
        """
        Adds a ``.toggle-div`` button.

        :param classes: Optional additional HTML classes.
        """
        assert self.collapsible_cid, "Please provide a CID for collapsible content"

        self._button_generator.addbutton(
            classes=classes,
        )

    def addcollapsiblediv(
            self,
            *,
            classes=(),  # type: typing.Sequence[str]
    ):  # type: (...) -> _HtmlDocumentType.NodeContext
        """
        Adds the div for the collapsible content.

        :param classes: Optional additional HTML classes.
        :return: HTML node context focused on the ``<div></div>`` node created.
        """
        assert self.collapsible_cid, "Please provide a CID for collapsible content"

        # Create the div.
        self._collapsible_div_ctx = self.html.addnode(
            "div",
            classes=[
                *classes,
                "collapsible",
                self.html.mkcsscompatibleclass(f"div={self.collapsible_cid}"),
            ],
        )

        # Finalize straight away if `default_state` is set.
        if self.default_state is not None:
            self.setstate(self.default_state)

        return self._collapsible_div_ctx

    def setstate(
            self,
            state,  # type: typing.Optional[_CollapsibleStateType]
    ):  # type: (...) -> None
        """
        Sets the default state for the collapsible content.

        :param state:
            State for the collapsible content.

            If ``None``, eventually sets the collapsible stuff as not-applicable.
        """
        from ._htmlgentypes import CollapsibleState

        assert self.collapsible_cid, "Please provide a CID for collapsible content"

        # Set toggle button `.expanded`/`.collapsed` class and text.
        self._button_generator.setstate(state)

        # Set collaspible `.expanded`/`.collapsed` class and visibility.
        # Memo: `addcollapsiblediv()` may have not been called.
        if self._collapsible_div_ctx is not None:
            with self._collapsible_div_ctx:
                if state is not None:
                    self.html.addclass(state)

                if state == CollapsibleState.EXPANDED:
                    self.html.addstyle("display: block")
                if state == CollapsibleState.COLLAPSED:
                    self.html.addstyle("display: none")

File: src/scenario/ui/_htmlgendivs.js
/*
 * Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @file
 * @brief Expandable/collapsible divs.
 */


/** @var {object} `scenario.divs` package. */
scenario.divs = {};


// Add 'click' event listeners on `a.toggle-div` buttons.
scenario.onLoad(() => {
    for (/** @var {HTMLElement} */ const _button of scenario.findNodes(document.body, "a.button.toggle-div")) {
        _button.addEventListener("click", (e) => {
            e.preventDefault();

            scenario.divs.toggle(_button);
        });
    }
});


/**
 * @brief Finds all collapsible divs in `focus` and applies `newState` on each.
 * @param {HTMLElement} focus Node which collapsible divs should be searched in.
 * @param {"expanded" | "collapsed"} newState New state to apply.
 * @param {((HTMLElement) => void) | undefined} buttonProcessed Callback for each `.toggle-div` button processed.
 * @returns {void}
 */
scenario.divs.findAndToggleAll = (focus, newState, {buttonProcessed} = {}) => {
    for (/** @var {HTMLElement} */ const _button of scenario.findNodes(focus, "a.button.toggle-div")) {
        // Process the `.toggle-div` button.
        scenario.divs.toggle(_button, {
            container: focus,  // Performance: restrict to `focus` at least.
            newState: newState,
        });

        // Call the `buttonProcessed()` callback if provided.
        if (buttonProcessed) {
            buttonProcessed(_button);
        }
    }
};


/**
 * @brief Shows or hides the `div.collapsible` attached with the given toggle button.
 * @param {HTMLElement} button Toggle button.
 * @param {string | null | undefined} divCid CID of the div to show/hide. Optional. Read from button classes if not provided.
 * @param {HTMLElement | null | undefined} container Container node for button and collapsible div. May be given, if known, for performance concerns.
 * @param {"expanded" | "collapsed" | undefined} newState Final state wanted. Switch state if not provided.
 * @returns {void}
 *
 * Calls `scenario.lists.checkToggleLi1()` on `button` (in case `button` stands for a `.li1` main list item).
 */
scenario.divs.toggle = (button, {divCid, container, newState} = {}) => {
    // Read CID from classes.
    if (! divCid) {
        divCid = scenario.getCid(button, "div");
    }

    // Ensure a candidate value for `container`.
    container = container || button.parentNode;

    // Find the collapsible div(s).
    /** @var {HTMLElement[]} */ const _collapsibleDivs = [];
    if (divCid) {
        /** @var {string} */ const _collapsibleDivSelector = `div.collapsible.div=${divCid}`;
        // Try from `container` first.
        if (container) {
            _collapsibleDivs.push(...scenario.findNodes(container, _collapsibleDivSelector));
        }
        // If not found try from `document.body`.
        if (! _collapsibleDivs.length) {
            _collapsibleDivs.push(...scenario.findNodes(document.body, _collapsibleDivSelector));
        }
    }

    // Determine old state from classes.
    /** @var {string | null} */ const _oldState = scenario.findOneClassOf(button, ["expanded", "collapsed"]);

    // Compute `newState` if not provided.
    if (! newState) {
        // Reverse status.
        switch (_oldState) {
            case "expanded": newState = "collapsed"; break;
            case "collapsed": newState = "expanded"; break;
        }
    }

    if (divCid && _collapsibleDivs.length && _oldState && newState) {
        console.debug(`Toggling div ${divCid}: ${_oldState} => ${newState}`);

        // Toggle button.
        scenario.buttons.setCollapsibleState(button, newState);

        // Collapsible div(s).
        for (/** @var {HTMLElement} */ const _collapsibleDiv of _collapsibleDivs) {
            // Remove old state and set new state on collapsible div.
            _collapsibleDiv.classList.remove("expanded");
            _collapsibleDiv.classList.remove("collapsed");
            _collapsibleDiv.classList.add(newState);

            // Adjust collapsible div visibility.
            switch (newState) {
                case "expanded": _collapsibleDiv.style.display = "block"; break;
                case "collapsed": _collapsibleDiv.style.display = "none"; break;
            }
        }

        // `.li` complementary actions.
        scenario.lists.checkToggleLi1(button, newState);
    }
};
